### PR TITLE
Kafka connector for CDC

### PR DESCRIPTION
**Pull request by Amogh-Bharadwaj**
This PR introduces Kafka connector for CDC where Kafka is a sync.
It also adds PLAINTEXT support for Kafka peers, along with SASL_SSL.

### CODE REVIEW

File: flow/activities/flowable.go [hunk 1]
```diff
@@ -82,15 +82,15 @@ func (a *FlowableActivity) CheckConnection(
 }
 
 // SetupMetadataTables implements IFlowable.SetupMetadataTables.
-func (a *FlowableActivity) SetupMetadataTables(ctx context.Context, config *protos.Peer) error {
+func (a *FlowableActivity) SetupMetadataTables(ctx context.Context, config *protos.Peer, jobName string) error {
```
Comment (iskakaushik) [inline flow/activities/flowable.go:85]: in `flow.proto` create a new message called `SetupMetadataTablesInput` that contains the `Peer` as well as `string jobName`. This makes backwards compatible changes easier in the future.

File: flow/connectors/kafka/kafka.go [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+	}
+
+	writer := &kafka.Writer{
+		Addr:                   kafka.TCP(brokers...),
+		AllowAutoTopicCreation: true,
+	}
+
+	if securityProtocol == "SASL_SSL" {
+		saslMechanism := plain.Mechanism{
+			Username: username,
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:54]: change locality of the variables.

File: flow/connectors/kafka/kafka.go [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+	controller, err := conn.Controller()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get controller to leader of Kafka cluster: %w", err)
+	}
+	leaderConn := net.JoinHostPort(controller.Host, strconv.Itoa(controller.Port))
+	controllerConn, err := dialer.Dial("tcp", leaderConn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get connection to leader controller: %w", err)
+	}
+	log.Println("Obtained controller connection.")
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:82]: use `Infoln` or `Infof`

File: flow/connectors/kafka/kafka.go [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+		return fmt.Errorf("error while closing connection to Kafka peer: %w", err)
+	}
+	return nil
+}
+
+func (c *KafkaConnector) ConnectionActive() bool {
+	if c == nil || c.connection == nil {
+		return false
+	}
+	_, err := c.connection.ReadPartitions()
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:112]: if there is an error, `log.Errorf` here.

File: flow/connectors/kafka/kafka.go [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+	return err != nil
+}
+
+func (c *KafkaConnector) NeedsSetupMetadataTables() bool {
+	return true
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:117]: check if the topic exists and return false if it does.

File: flow/connectors/kafka/kafka.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+func (c *KafkaConnector) SetupMetadataTables(jobName string) error {
+	metadataTopicName := "peerdb_" + jobName
+	createErr := c.connection.CreateTopics(kafka.TopicConfig{
+		Topic:             metadataTopicName,
+		NumPartitions:     1,
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:124]: make this more reliable, pick good defaults to start with.

File: flow/connectors/kafka/kafka.go [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+	})
+
+	if createErr != nil {
+		return fmt.
+			Errorf("failed to create Kafka topic: %w", createErr)
+	}
+	return nil
+}
+
+func (c *KafkaConnector) GetLastOffset(jobName string) (*protos.LastSyncState, error) {
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:135]: add a couple of comments, 1 near the place where we set num partitions for this topic to 1 as well.

File: flow/connectors/kafka/kafka.go [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+	if last == 0 {
+		log.Warnf("Assuming zero offset means no sync has happened for job %s, returning nil", jobName)
+		return nil, nil
+	}
+	setErr := offsetReader.SetOffset(last - 1)
+	if setErr != nil {
+		return nil, fmt.Errorf("failed to set offset to the last: %w", setErr)
+	}
+	lastMessage, err := offsetReader.ReadMessage(c.ctx)
+
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:157]: remove this newline

File: flow/connectors/kafka/kafka.go [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+func (c *KafkaConnector) GetLastOffset(jobName string) (*protos.LastSyncState, error) {
+	metadataTopicName := "peerdb_" + jobName
+	conn, err := c.dialer.DialLeader(c.ctx, "tcp", c.connection.RemoteAddr().String(), metadataTopicName, 0)
+	if err != nil {
+		return nil, fmt.Errorf("unable to dial metadata table: %w", err)
+	}
+	last, err := conn.ReadLastOffset()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get last offset of metadata table: %w", err)
+	}
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:144]: after each error handling block add a new line.

File: flow/connectors/kafka/kafka.go [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+
+	if err != nil {
+		log.Warnf("error fetching last offset: %s", err)
+		return nil, nil
+	}
+
+	lastCheckpoint, integerParseErr := strconv.ParseInt(string(lastMessage.Value), 10, 64)
+	if integerParseErr != nil {
+		log.Warnf("error converting checkpoint string to int64: %s", integerParseErr)
+		return nil, nil
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:166]: shouldn't this be an error?

File: flow/connectors/kafka/kafka.go [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+	if err := offsetReader.Close(); err != nil {
+		return nil, fmt.Errorf("failed to close last offset reader: %w", err)
+	}
+
+	return &protos.LastSyncState{
+		Checkpoint: lastCheckpoint,
+	}, nil
+}
+
+func (c *KafkaConnector) GetLastSyncBatchID(jobName string) (int64, error) {
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:178]: TODO

File: flow/connectors/kafka/kafka.go [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+			Errorf("failed to create Kafka topic: %w", createErr)
+	}
+	return &protos.SetupNormalizedTableOutput{
+		TableIdentifier: destinationTopicName,
+		AlreadyExists:   false,
+	}, nil
+}
+
+func (c *KafkaConnector) InitializeTableSchema(req map[string]*protos.TableSchema) error {
+	return nil
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:212]: log.Info saying that this is expected.

File: flow/connectors/kafka/kafka.go [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+func (c *KafkaConnector) PullRecords(req *model.PullRecordsRequest) (*model.RecordBatch, error) {
+	log.Errorf("panicking at call to PullRecords for Kafka flow connector")
+	panic("PullRecords is not implemented for the Kafka flow connector")
+}
+
+func (c *KafkaConnector) SyncRecords(req *model.SyncRecordsRequest) (*model.SyncResponse, error) {
+	var destinationTopicName string
+	if len(req.Records.Records) > 0 {
+		destinationTopicName = "peerdb_final_" + req.Records.Records[0].GetTableName()
+	}
```
Comment (iskakaushik) [inline flow/connectors/kafka/kafka.go:224]: else return  right here.

File: flow/connectors/kafka/qrep.go [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+func (c *KafkaConnector) SyncQRepRecords(
+	config *protos.QRepConfig,
+	partition *protos.QRepPartition,
+	records *model.QRecordBatch,
+) (int, error) {
+	panic("not implemented")
+}
+
+func (c *KafkaConnector) ConsolidateQRepPartitions(config *protos.QRepConfig) error {
```
Comment (iskakaushik) [inline flow/connectors/kafka/qrep.go:28]: make this also panic, for all the panics in this class say `kafka doesn't yet support query replication`.

**Comment by Amogh-Bharadwaj** [comment]
Closed in favour of #815

**Pull request closed without merging**